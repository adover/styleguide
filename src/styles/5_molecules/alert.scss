@import '../1_settings/_config';
@import '../2_tools/mixins/apply-spacing';
@import '../2_tools/mixins/display-flex';
@import '../4_atoms/icon';
@import '../4_atoms/section';
@import '../5_molecules/modal';
@import '../5_molecules/slide-panel';

/// @group molecules

$alert-font-size: map-deep-get($config, sizes, paragraph, small);
$alert-max-height: 40rem;

$border-highlight-width: $spacing-2;

$alert-icon: (
	size: $spacing-8,
	margin-right: $spacing-2,
);

///
/// @name alert-init
/// @group initialiser
@mixin alert-init($iconClass: alert-icon, $animation: true, $fullWidth: false) {
	@include apply-spacing($size: default);
	@include section-margin($size: medium);
	@include slide-panel-init();
	@include display-flex($direction: row, $flex: 0, $flex-wrap: wrap);

	@include media-query($size: mobile) {
		align-items: center;
		justify-content: space-between;
	}

	@if $animation {
		animation: shakeAndShow map-get($transitionSpeed, medium) cubic-bezier(0.36, 0.07, 0.19, 0.97) both;
	}

	@if $fullWidth {
		width: 100%;
	}

	// Use this delay to animate the container height
	animation-delay: map-get($transitionSpeed, default);
	backface-visibility: hidden;
	border: 1px solid transparent;
	border-left: $border-highlight-width solid;
	border-radius: $radius-default;
	font-size: $alert-font-size;
	margin-left: auto;
	margin-right: auto;
	max-height: $alert-max-height;
	max-width: if($fullWidth, none, map-deep-get($config, page, medium-width));
	position: relative;
	transform: translate3d(0, 0, 0);

	@each $alert-type, $type-list in map-deep-get($config, alerts) {
		&[data-type='#{$alert-type}'] {
			background: map-get($type-list, background-color);

			@if map-get($type-list, border) == 'all' {
				border-color: map-get($type-list, accent-color);
			}

 @else {
				border-left-color: map-get($type-list, accent-color);
			}
		}
	}

	span {
		display: block;
	}
}

@mixin alert-can-close() {
	padding-right: ($modal-close-size * 2) !important;
}

@mixin alert-content() {
	flex: 1;
	margin: auto;
}

@mixin alert-close() {
	@include modal-close();
}

@mixin alert-label() {
	font-weight: 700;
}

@mixin alert-icon() {
	@include icon-init();

	flex: 0 0 $svg-size;
	margin-right: map-get($alert-icon, margin-right);

	@include media-query($size: tablet-portrait) {
		margin-left: -#{map-get($alert-icon, margin-right)};
	}
}
