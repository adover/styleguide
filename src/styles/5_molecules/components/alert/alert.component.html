<p class="cdAlert"
  [ngClass]="'cdAlert--' + alertType"
  *ngIf="shouldDisplayAlert">
  <!--
  <mat-icon [svgIcon]="icon"
    class="cdAlert-icon"></mat-icon>
-->
  <span class="cdAlert-message"
    role="alert">
    <ng-content></ng-content>
  </span>

  <button aria-label="Dismiss alert"
    class="cdAlert-dismissButton"
    *ngIf="canDismiss"
    (click)="closeAlert()">
<!--    <mat-icon svgIcon="cross-ui"
      [inline]="true"></mat-icon>
  -->
  </button>
</p>
