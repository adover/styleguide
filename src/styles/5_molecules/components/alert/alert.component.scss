@import '../../alert';
@import '../../../1_settings/transitions';
@import '../../../2_tools/functions/build-transition';

:host {
	animation: height map-get($transitionSpeed, medium) cubic-bezier(0.36, 0.07, 0.19, 0.97) forwards,
		opacity map-get($transitionSpeed, medium) cubic-bezier(0.36, 0.07, 0.19, 0.97) forwards;
	display: block;
	max-height: 0;
	overflow: hidden;

	&[cananimate='false'] {
		animation: none;
		max-height: 9999px;

		.alert {
			animation: none;
		}
	}
}

.alert {
	&[data-fullwidth='true'] {
		@include alert-init($fullWidth: true);
	}

	&:not([data-fullwidth='true']) {
		@include alert-init();
	}

	&[data-withmargin='false'] {
		margin-bottom: 0;
	}

	&--candismiss {
		@include alert-can-close();
	}

	.alert-icon {
		@include alert-icon();
	}

	&-closeBtn {
		@include alert-close();
	}

	&-label {
		@include alert-label();
	}

	&-cta {
		.alert & {
			@include section-margin($size: default, $side: top);

			@include media-query($size: mobile) {
				@include section-margin($size: default, $side: left);

				flex-basis: 0;
				margin-top: 0;
			}

			flex-basis: 100%;
		}
	}

	&-content {
		.alert & {
			@include alert-content();
		}

		&::ng-deep {
			a {
				text-decoration: underline;

				&:hover {
					text-decoration: none;
				}
			}
		}
	}

	&-description:not(:last-child) {
		@include section-margin($size: default);
	}
}
