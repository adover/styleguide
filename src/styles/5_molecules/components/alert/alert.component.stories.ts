import { boolean, select, text, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/angular';
import { ButtonComponent } from '../../../4_atoms/components';
import {
	SvgIconComponent,
	SvgDefinitionsComponent,
	AlertComponent,
	APP_SETTINGS_PROVIDER,
} from '@woolworthsnz/styleguide';

storiesOf('Molecules| Alert', module)
	.addDecorator(withKnobs)
	.add('default', () => {
		const canDismiss = boolean('Dismissable?', true);
		const alertTypes = {
			info: 'info',
			success: 'success',
			warning: 'warning',
			error: 'error',
		};
		const alertType = select('Alert Type', alertTypes, 'error');
		const alertTitle = text('Alert Title', undefined);
		const alertDescription = text('Alert Description', undefined);
		const showCta = boolean('Show CTA on right?', true);

		return {
			moduleMetadata: {
				declarations: [
					APP_SETTINGS_PROVIDER,
					AlertComponent,
					SvgIconComponent,
					SvgDefinitionsComponent,
					ButtonComponent,
				],
			},
			template: `
        <cdx-alert [candismiss]="canDismiss" [type]="alertType" [title]="alertTitle" [description]="alertDescription" [showCta]="showCta">
          <button cdxButton fillStyle="secondary" cta>Click click click</button>
        </cdx-alert>
        <cdx-svg-definitions></cdx-svg-definitions>`,
			props: {
				alertType,
				canDismiss,
				alertDescription,
				alertTitle,
				showCta,
			},
		};
	});
