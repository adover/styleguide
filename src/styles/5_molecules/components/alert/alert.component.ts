import { SVGIcon } from '../../../4_atoms/components/svg-definitions/SVGIcon';
import { ValidationError } from '../../../model/validationError';
import { AlertType } from '../../../ui-models';
import {
	ChangeDetectionStrategy,
	ChangeDetectorRef,
	Component,
	ElementRef,
	HostBinding,
	Input,
	OnInit,
	ViewChild,
} from '@angular/core';

@Component({
	selector: 'cdx-alert, [cdxAlert]',
	template: `
		<div
			class="alert"
			[attr.data-withmargin]="withMargin"
			[attr.data-fullwidth]="fullwidth"
			[attr.labelledBy]="alertLabelId"
			[attr.describedBy]="alertDescriptionId"
			[attr.data-type]="type"
			[attr.role]="role"
			[class.alert--candismiss]="candismiss"
			*ngIf="showAlert"
		>
			<cdx-svg-icon class="alert-icon" [name]="icon" [fill]="type" *ngIf="icon"></cdx-svg-icon>

			<div class="alert-content">
				<span #content class="alert-label" id="{{ alertLabelId }}">
					<ng-container *ngIf="title; else fallback">{{ title }} </ng-container>
					<ng-template #fallback>{{ fallbackMessage }}</ng-template>
				</span>

				<div *ngIf="description" id="{{ alertDescriptionId }}" class="alert-description">
					<span *ngFor="let item of description">
						<ng-container *ngIf="item?.message; else descriptionTemplate">
							<ng-container *ngIf="item?.field">{{ item.field }}: </ng-container>{{ item?.message }}
						</ng-container>

						<ng-template #descriptionTemplate>
							<span [innerHTML]="item"></span>
						</ng-template>
					</span>
				</div>

				<ng-content select="[alertLink]"></ng-content>
			</div>

			<div *ngIf="showCta" class="alert-cta">
				<ng-content select="[cta]"></ng-content>
			</div>
			<button aria-label="Dismiss alert" class="alert-closeBtn" *ngIf="candismiss" (click)="closeAlert()"></button>
		</div>
	`,
	styleUrls: ['./alert.component.scss'],

	changeDetection: ChangeDetectionStrategy.OnPush,
})
export class AlertComponent implements OnInit {
	_alertType: AlertType = AlertType.info;
	_description: string[] | ValidationError[];

	_icon: SVGIcon | false = SVGIcon.Alert;

	dismissed = false;

	get showAlert() {
		return !this.dismissed;
	}

	get role() {
		// https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_alert_role#Notes
		return this.candismiss ? 'alertdialog' : 'alert';
	}

	get type() {
		return this._alertType;
	}

	@Input() set type(alertType: AlertType) {
		this._alertType = alertType ? alertType : AlertType.error;
	}

	get icon() {
		return this._icon;
	}

	@Input() set icon(icon: SVGIcon | false) {
		this._icon = icon;
	}

	@Input() showCta = false;

	@Input() title: string;

	@HostBinding('attr.cananimate')
	@Input()
	animate = true;

	// not camelCase due to using this as a custom element
	@Input()
	fullwidth = false;

	get description() {
		return this._description;
	}

	@Input() set description(description: string | string[] | ValidationError[]) {
		if (!description) return;
		this._description = [].concat(description);
	}

	@Input()
	candismiss: Boolean = false;

	@Input() withMargin = true;

	@ViewChild('content', { static: true }) content: ElementRef;

	alertLabelId = 'alertLabel';
	alertDescriptionId: string;

	alertTypes = AlertType;
	fallbackMessage =
		'Sorry an error has occurred. If you continue to receive these errors please contact our Customer Service Team on 0800 40 40 40';

	constructor(private element: ElementRef, private cdr: ChangeDetectorRef) {}

	ngOnInit(): void {
		this.alertDescriptionId = this.description && this.description.length ? 'alertDescription' : undefined;
	}

	closeAlert() {
		this.dismissed = true;

		this.cdr.markForCheck();
	}

	hasMessage = () => {
		return (
			this.content.nativeElement.childNodes.length > 0 &&
			Boolean(this.content.nativeElement.childNodes[0].textContent.trim())
		);
	};
}
