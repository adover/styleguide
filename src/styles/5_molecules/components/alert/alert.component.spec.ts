import { Component, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { AlertComponent } from './alert.component';
import { AlertType } from '../../../ui-models';

@Component({
	selector: `cdx-host`,
	template: `
		<cdx-alert [candismiss]="candismiss" [type]="alertType" [title]="content" [description]="description"></cdx-alert>
	`,
})
class DummyComponent {
	candismiss: boolean;
	alertType: AlertType;
	content: string;
	description: string | string[];
}

describe('AlertComponent', () => {
	let dummyComponent: DummyComponent;
	let fixture: ComponentFixture<DummyComponent>;

	beforeEach(async(() => {
		TestBed.configureTestingModule({
			imports: [],
			declarations: [AlertComponent, DummyComponent],
			schemas: [CUSTOM_ELEMENTS_SCHEMA],
		}).compileComponents();
	}));

	beforeEach(() => {
		fixture = TestBed.createComponent(DummyComponent);
		dummyComponent = fixture.componentInstance;

		fixture.detectChanges();
	});

	it('should create', () => {
		expect(dummyComponent).toBeTruthy();
	});

	it('sets the correct dataType based on the alertType input', () => {
		dummyComponent.alertType = AlertType.info;

		fixture.detectChanges();
		expect(fixture).toMatchSnapshot();
	});

	it('displays a fallback message if ng content is empty', () => {
		expect(fixture.debugElement.query(By.css('#alertLabel')).nativeElement.textContent.trim()).toBe(
			'Sorry an error has occurred. If you continue to receive these errors please contact our Customer Service Team on 0800 40 40 40'
		);
	});

	xit('Calls closeAlert when the close button is clicked on', () => {
		// TODO: Method not implemented yet
	});

	it('Updates the host class if it can be closed', () => {
		dummyComponent.candismiss = true;

		fixture.detectChanges();
		expect(fixture).toMatchSnapshot();
	});

	it('Has a role of alertdialog if it can dismiss', () => {
		dummyComponent.candismiss = true;

		fixture.detectChanges();
		expect(fixture.debugElement.query(By.css('.alert')).nativeElement.getAttribute('role')).toBe('alertdialog');
	});

	it('Has a role of alert if it cannot dismiss', () => {
		dummyComponent.candismiss = false;

		fixture.detectChanges();
		expect(fixture.debugElement.query(By.css('.alert')).nativeElement.getAttribute('role')).toBe('alert');
	});
});
