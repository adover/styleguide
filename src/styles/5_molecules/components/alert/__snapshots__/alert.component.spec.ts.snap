// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertComponent Updates the host class if it can be closed 1`] = `
<cdx-host
  __ngContext__={[Function LRootView]}
  candismiss={[Function Boolean]}
>
  <cdx-alert
    cananimate="true"
  >
    <div
      class="alert alert--candismiss"
      data-fullwidth="false"
      data-type="error"
      data-withmargin="true"
      labelledby="alertLabel"
      role="alertdialog"
    >
      <cdx-svg-icon
        class="alert-icon"
      />
      
      <div
        class="alert-content"
      >
        <span
          class="alert-label"
          id="alertLabel"
        >
          Sorry an error has occurred. If you continue to receive these errors please contact our Customer Service Team on 0800 40 40 40
          
          
        </span>
        
      </div>
      
      <button
        aria-label="Dismiss alert"
        class="alert-closeBtn"
      />
      
    </div>
    
  </cdx-alert>
</cdx-host>
`;

exports[`AlertComponent sets the correct dataType based on the alertType input 1`] = `
<cdx-host
  __ngContext__={[Function LRootView]}
  alertType={[Function String]}
>
  <cdx-alert
    cananimate="true"
  >
    <div
      class="alert"
      data-fullwidth="false"
      data-type="info"
      data-withmargin="true"
      labelledby="alertLabel"
      role="alert"
    >
      <cdx-svg-icon
        class="alert-icon"
      />
      
      <div
        class="alert-content"
      >
        <span
          class="alert-label"
          id="alertLabel"
        >
          Sorry an error has occurred. If you continue to receive these errors please contact our Customer Service Team on 0800 40 40 40
          
          
        </span>
        
      </div>
      
      
    </div>
    
  </cdx-alert>
</cdx-host>
`;
